### ISSUE TITLE
[HOLD for payment 2023-05-16] [$1000] Web - Chat - Inconsistent parsing on occurrence of more than two consecutive styling characters

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17665]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open [staging.new.expensify.com](http://staging.new.expensify.com/)
2. Send this message : *** (triple asterisks)
3. Send this message : ___  (triple underscores)
4. Examine the difference of the parsing between the messages

## Expected Result:
Both of the messages should be parsed the same, either all characters styling isn't parsed or the middle character got parsed

## Actual Result:
While the bold characters (asterisks) isn't parsed, the italic characters (underscores) got parsed and only leave one underscore after parsed

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.1.3

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/233133090-5616da89-20ef-4e4c-b358-3e110e0d8a58.mp4

https://user-images.githubusercontent.com/93399543/233133119-ad030e7e-14c2-4702-a2ad-4ee4e1070a0e.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** @kerupuksambel

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1681867484400789

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ab464dbcaa9a04d3</li>
        <li>Upwork Job ID: 1648736676150435840</li>
        <li>Last Price Increase: 2023-04-26</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @zanyrenney (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by zanyrenney** [comment]
Going to check in the channel.

**Comment by zanyrenney** [comment]
https://expensify.slack.com/archives/C01SKUP7QR0/p1681921007844539?thread_ts=1681872394.446279&cid=C01SKUP7QR0

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01ab464dbcaa9a04d3

**Comment by MelvinBot** [comment]
Current assignee @zanyrenney is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @hayata-suenaga (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dukenv0307** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
While the bold characters (asterisks) isn't parsed, the italic characters (underscores) got parsed and only leave one underscore after parsed.

> either all characters styling isn't parsed or the middle character got parsed.

We should take about what we want to do here first since there're 2 ways mentioned here. If we try some other popular markdown editor (like Github itself), we'll see that the styling character itself will not be included in the final styled text. For example if we try `test ___` (3 underscores) on Github, it will show as is without styling the 3 underscores.

So I think we should do the same here. It also doesn't make sense to style the middle `_` since `_` by itself does not have any italic style.

### What is the root cause of that problem?

> While the bold characters (asterisks) isn't parsed

This is because in https://github.com/Expensify/expensify-common/blob/07db321d9dffbc66a56d7d7bf42ba1cbf3df216e/lib/ExpensiMark.js#L143, we're excluding the `*` if it's inside the `* *` pair (see the `([^\s*]|\s(?!\*))` part).

While in https://github.com/Expensify/expensify-common/blob/07db321d9dffbc66a56d7d7bf42ba1cbf3df216e/lib/ExpensiMark.js#L135, we're not doing that, we still allow the `_` character. That's why the problem occurs for the `_`
 
### What changes do you think we should make in order to solve the problem?

We need to update this regex https://github.com/Expensify/expensify-common/blob/07db321d9dffbc66a56d7d7bf42ba1cbf3df216e/lib/ExpensiMark.js#L135 to not allow the lonely `_` character inside the `_`, `_` pair. We can do the same as we already do for the `*` as explained above.

### What alternative solutions did you explore? (Optional)
NA

**Comment by romulo114** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Consecutive styling characters are not parsed consistently.

### What is the root cause of that problem?
You can see the `regex` property of `bold` rule [here](https://github.com/Expensify/expensify-common/blob/be625f55058111793bbcfe3f64788ea0a8f9d705/lib/ExpensiMark.js#L149-L151)
```
    name: 'bold',
    regex: /\B\*((?=\S)(([^\s*]|\s(?!\*))+?))\*\B(?![^<]*(<\/pre>|<\/code>|<\/a>))/g,
    replacement: (match, g1) => (g1.includes('<pre>') ? match : `<strong>${g1}</strong>`),
```
The main reason of this issue is `[^\s*]`, which prevents `*` characters inside the ** from being matched.

### What changes do you think we should make in order to solve the problem?
We need to remove the * from the above regex like below.
```
    regex: /\B\*((?=\S)(([^\s]|\s(?!\*))+?))\*\B(?![^<]*(<\/pre>|<\/code>|<\/a>))/g,
```

### Result

https://user-images.githubusercontent.com/117526965/233333448-7d7fc5b8-8351-4dc6-81ba-8ad4097d0bf3.mp4


### What alternative solutions did you explore? (Optional)
N/A

**Comment by hayata-suenaga** [comment]
I believe both should not be parsed

**Comment by zanyrenney** [comment]
Heading OOO, reapplying the Bug0 label for active management.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @alexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by hasebsiddiqui** [comment]
I believe my [proposal](https://github.com/Expensify/App/issues/17665#issuecomment-1515292795) is the right one for this job. As it is the first to identify the problem and provide the solution. All other solutions are similar to my proposal.

**Comment by dukenv0307** [comment]
thanks @parasharrajat for the review!

I've updated [my proposal](https://github.com/Expensify/App/issues/17665#issuecomment-1516037067) to remove the use of lookbehind.

**Comment by hasebsiddiqui** [comment]
I was following guidelines provided on upwork by Expensify where it's written then `You are not required to submit the final solution or code along with your proposal.`. I thought I have already explained the problem and provided the video evidence in my proposal after solving the problem without providing the final code. 
<img width="990" alt="image" src="https://user-images.githubusercontent.com/43096403/233950107-793f0be7-1797-46e2-b7f7-d9f06b690941.png">

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by alexpensify** [comment]
Waiting on proposals here

**Comment by alexpensify** [comment]
Waiting on proposals here

**Comment by romulo114** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17665#issuecomment-1516061868)

**Comment by alexpensify** [comment]
@parasharrajat - when you get a chance, may you please review the updated proposal? Thanks!

**Comment by hasebsiddiqui** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/17665#issuecomment-1515292795)

I updated on 26 - April. Forgot to comment here.

**Comment by hayata-suenaga** [comment]
@parasharrajat thank you for your reviews let's go with @dukenv0307's proposal

**Comment by MelvinBot** [comment]
📣 @dukenv0307 You have been assigned to this job by @hayata-suenaga!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by MelvinBot** [comment]
@alexpensify @parasharrajat @hayata-suenaga @dukenv0307 this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by alexpensify** [comment]
Melvin, the PR is in review! 👏🏼

**Comment by alexpensify** [comment]
PR has been merged!

**Comment by alexpensify** [comment]
Thank you, I updated my comment.

**Comment by alexpensify** [comment]
The new PR is in the works!

**Comment by hayata-suenaga** [comment]
@alexpensify can you check @parasharrajat's comments?

**Comment by alexpensify** [comment]
Thank you @parasharrajat for the work here. I'm aware of the two PRs and I'll review the payment breakout later this week since payment is not due until next week.

**Comment by alexpensify** [comment]
Still on hold for payment.

**Comment by alexpensify** [comment]
I've started an internal discussion to confirm the payments.

**Comment by alexpensify** [comment]
I've reviewed the two PRs with the team. We have confirmed that it is common that some GitHubts like this one could require two PRs. One PR to make the update in expensify-common and then another to bring the updated expensify-common version in App. With that said, the bonus will not apply here. The merge was past the 3-business day mark. With that said, this GH has the following payouts. 
 
Issue reporter: $250 - @kerupuksambel
Contributor: $1,000 - @dukenv0307
Contributor+: $1,000- @parasharrajat

If anyone has feedback please share. If no response, tomorrow, I’ll complete the next steps in Upwork. Thank you!

**Comment by dylanexpensify** [comment]
@kerupuksambel, @dukenv0307, and @parasharrajat please apply [here](https://www.upwork.com/jobs/~01ab464dbcaa9a04d3)!

**Comment by dylanexpensify** [comment]
@dukenv0307 please apply! @parasharrajat @kerupuksambel sent offers!

**Comment by dukenv0307** [comment]
@dylanexpensify applied, thank you!

**Comment by dylanexpensify** [comment]
@dukenv0307 offer sent!

**Comment by dylanexpensify** [comment]
Done!

**Comment by alexpensify** [comment]
Thanks for the assist here @dylanexpensify!
